plugins/microsoft-graph-authz/src/main/java/org/elasticsearch/xpack/security/authz/microsoft/MicrosoftGraphAuthzRealm.java

```diff
@@ -43,8 +90,91 @@ public void authenticate(AuthenticationToken token, ActionListener<Authenticatio
     }
 
     @Override
-    public void lookupUser(String username, ActionListener<User> listener) {
-        logger.info("Microsoft Graph Authz not yet implemented, returning empty roles for [{}]", username);
-        listener.onResponse(new User(username));
+    public void lookupUser(String principal, ActionListener<User> listener) {
+        try {
+            // TODO probably want to do this once rather than every time
+            final var client = buildClient();
+            final var userProperties = sdkFetchUserProperties(client, principal);
+            final var groups = sdkFetchGroupMembership(client, principal);
+
+            // TODO confirm we don't need any other fields
+            final var userData = new UserRoleMapper.UserData(principal, null, groups, Map.of(), config);
+
+            roleMapper.resolveRoles(userData, listener.delegateFailureAndWrap((l, roles) -> {
+                final var user = new User(
+                    principal,
+                    roles.toArray(Strings.EMPTY_ARRAY),
+                    userProperties.v1(),
+                    userProperties.v2(),
+                    Map.of(),
+                    true
+                );
+                logger.trace("Entra ID user {}", user);
+                l.onResponse(user);
+            }));
+        } catch (Exception e) {
+            // TODO logging etc
+            listener.onFailure(e);
+        }
+    }
+
+    private GraphServiceClient buildClient() {
+        logger.trace("building client");
+        final var credentialProvider = new ClientSecretCredentialBuilder().clientId(
+            config.getSetting(MicrosoftGraphAuthzRealmSettings.CLIENT_ID)
+        )
+            .clientSecret(clientSecret.toString())
+            .tenantId(config.getSetting(MicrosoftGraphAuthzRealmSettings.TENANT_ID))
+            .authorityHost(config.getSetting(MicrosoftGraphAuthzRealmSettings.ACCESS_TOKEN_HOST))
+            // TODO this is necessary for tests, but we probably want this enabled in prod
+            .disableInstanceDiscovery()
```

Comment:
to be clear: when instance discovery is on, the lib expects the host to be a known MS host, or will call MS to attempt to verify the host - this obviously doesn't work when we're trying to connect to localhost